modules/apps/headless/headless-builder/headless-builder-impl/src/main/java/com/liferay/headless/builder/internal/model/listener/HeadlessBuilderObjectEntryModelListener.java

```diff
@@ -0,0 +1,154 @@
+
+import java.io.Serializable;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Sergio Jiménez del Coso
+ */
+@Component(service = ModelListener.class)
+public class HeadlessBuilderObjectEntryModelListener
+	extends BaseModelListener<ObjectEntry> {
+
+	@Override
+	public void onAfterCreate(ObjectEntry objectEntry)
+		throws ModelListenerException {
+
+		_schedulePublication(objectEntry);
+	}
+
+	@Override
+	public void onAfterUpdate(
+			ObjectEntry originalObjectEntry, ObjectEntry objectEntry)
+		throws ModelListenerException {
+
+		_schedulePublication(objectEntry);
+	}
+
+	private long _getAPIApplicationId(ObjectEntry objectEntry) {
+		if (_isObjectDefinition("L_API_APPLICATION", objectEntry)) {
+			return objectEntry.getObjectEntryId();
+		}
+		else if (_isObjectDefinition("L_API_SCHEMA", objectEntry)) {
+			Map<String, Serializable> values = objectEntry.getValues();
+
+			return (long)values.get(
+				"r_apiApplicationToAPISchemas_c_apiApplicationId");
+		}
+		else if (_isObjectDefinition("L_API_ENDPOINT", objectEntry)) {
+			Map<String, Serializable> values = objectEntry.getValues();
+
+			return (long)values.get(
+				"r_apiApplicationToAPIEndpoints_c_apiApplicationId");
+		}
+
+		throw new UnsupportedOperationException("Invalid object definition");
+	}
+
+	private boolean _isObjectDefinition(
+		String objectDefinitionExternalReferenceCode, ObjectEntry objectEntry) {
+
+		ObjectDefinition objectDefinition =
+			_objectDefinitionLocalService.fetchObjectDefinition(
+				objectEntry.getObjectDefinitionId());
+
+		return Objects.equals(
+			objectDefinition.getExternalReferenceCode(),
+			objectDefinitionExternalReferenceCode);
+	}
+
+	private void _schedulePublication(ObjectEntry objectEntry) {
+		ObjectDefinition objectDefinition =
+			_objectDefinitionLocalService.fetchObjectDefinition(
+				objectEntry.getObjectDefinitionId());
+
+		if (StringUtil.startsWith(
+				objectDefinition.getExternalReferenceCode(), "L_API_")) {
+
+			long apiApplicationId = _getAPIApplicationId(objectEntry);
+
+			_pendingAPIApplications.add(apiApplicationId);
+
+			TransactionCommitCallbackUtil.registerCallback(
+				() -> {
+					if (_pendingAPIApplications.contains(apiApplicationId)) {
+						ObjectEntry apiApplicationObjectEntry =
+							_objectEntryLocalService.getObjectEntry(
+								apiApplicationId);
+
+						Map<String, Serializable> values =
+							apiApplicationObjectEntry.getValues();
+
+						APIApplication apiApplication =
+							_apiApplicationProvider.fetchAPIApplication(
+								(String)values.get("baseURL"),
+								apiApplicationObjectEntry.getCompanyId());
+
+						if (StringUtil.equals(
+								(String)values.get("applicationStatus"),
+								"published")) {
+
+							_apiApplicationPublisher.unpublish(apiApplication);
```

Comment:
The `_apiApplicationPublisher.unpublish(apiApplication);` could be moved out of the if/else clause as it is being always executed

Reply:
@sergiojimcos ☝️

Reply:
https://github.com/liferay-headless/liferay-portal/pull/1321/commits/78cfb2a2138d38eb41e317cdf53caaeb2f444c5c